### PR TITLE
Spring Boot timezones

### PR DESCRIPTION
**Pull request by ranjanih**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [ ] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [ ] the article explains *why* and not only *how*
- [ ] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [ ] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [ ] the text is conversational ("we" instead of "you", easy language)
- [ ] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [ ] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [ ] the text uses active voice instead of passive voice
- [ ] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [ ] file names and variables are highlighted as `code`

## Structure
- [ ] the article has a short intro before the first headline
- [ ] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [ ] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [ ] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [ ] code examples are introduced with a ":" and explained below the code example
- [ ] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+### Missing Timezone information
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/date.JPG" %}}
+
+- `java.util.Date` represents an instant in time. 
+- Also, it has no timezone information. So, it considers a **default system timezone** which could differ from location to location.
+For instance, if a person runs this test from another country, he might see a different date and time derived from the given milliseconds.
+
+### Custom Date creation
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/customDate.JPG" %}}
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:43]: Instead of a screenshot of code and console output, can you add the code and the console output into the article itself? That's better readable and more inclusive.

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+
+- Creating a custom date with this API is very inconvenient. Firstly, the year starts with 1900, hence we must
+subtract 1900 so that the right year is considered.
+- Also to derive the months we need to use indexes 0-11. In this example, to create a date in August we would use 7 and not 8.
+
+### Mutable Classes
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/mutableDate.JPG" %}}
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:51]: Same with this screenshot.

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+
+- Immutability is a key concept that ensures that java objects are thread-safe and concurrent access does not lead to an inconsistent state.
+- The Date API **allows mutable methods** such as `setHours`, `setMinutes`, `setDate`. Therefore, it becomes the **responsibility
+of the developer to clone the object before use**.
+- Similarly, the Calendar object also has setter methods `setTimeZone`, `add` which allows an object to be modified.
+
+### Format Dates
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/dateFormat.JPG" %}}
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:60]: Same here (and for all other screenshots that can be converted into code, I'm not going to point them all out).

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+- Creating a custom date with this API is very inconvenient. Firstly, the year starts with 1900, hence we must
+subtract 1900 so that the right year is considered.
+- Also to derive the months we need to use indexes 0-11. In this example, to create a date in August we would use 7 and not 8.
+
+### Mutable Classes
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/mutableDate.JPG" %}}
+
+- Immutability is a key concept that ensures that java objects are thread-safe and concurrent access does not lead to an inconsistent state.
+- The Date API **allows mutable methods** such as `setHours`, `setMinutes`, `setDate`. Therefore, it becomes the **responsibility
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:54]: Please check that all references to method names in text contain parentheses (i.e. `setHours()`, `setMinutes()`). I've done it here, but not in other places in this article.

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+As seen in the example above, there are various flaws in this process:
+- The Date API itself does not store any formatting information. Therefore, we need to use it in combination with the `SimpleDateFormat`.
+- The `SimpleDateFormat` class is not thread-safe so it **cannot be used in multithreaded applications without proper synchronization**.
+- As the Date API **does not have timezone information**, we have to use the `Calendar` class. However, it cannot be formatted, 
+so we extract date from `Calendar` for formatting.
+
+#### SQL Dates
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/sqlDates.JPG" %}}
+
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:72]: Maybe introduce with a sentence what `java.sql.Date` is used for?

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+- The API demonstrates poor design choice as `java.sql.Date`, `java.sql.Time` and `java.sql.Timestamp` all extend `java.util.Date` class.
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:73]: Why are those bad design choices?

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+{{% /info %}}
+
+## Java8 DateTime API
+
+The [DateTime API](https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html) is heavily influenced by the [Jodatime](https://www.joda.org/joda-time/) library which was the defacto standard prior to Java8.
+In this section, we will look at some commonly used date-time classes and its corresponding operations in this section.
+
+### LocalDate
+
+**[java.time.LocalDate](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html)** is an immutable date object that does not store time or timezone information. However, we can pass the `java.time.ZoneId` object to get the local date in a particular timezone.
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:89]: > is an immutable date object that does not store time or timezone information

Well, what _does_ it store, then? :)

(I know it stores a date, but it doesn't hurt to be explicit here)

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+
+Sample conversion examples: 
+
+````java
+
+        LocalDate today = LocalDate.now();
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:95]: This and the following code snippets are worth gold for people who are searching for conversions between date objects!

However, it's hard to follow all the `System.out.println()`s. Can you convert the code examples to unit tests with assertions? This way, the reader sees an assertions like `assertThat(today).hasDayOfMonth(12)`. If you use assertJ as the assertion library (like in the example) the tests become very readable and the reader gets the most value (and learns that assertJ is a cool library to test dates).

Also, you wouldn't need to show any console output at all, because the tests are self-explanatory.

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+## Compatibility with the legacy API
+
+As a part of the Date/Time API, **methods have been introduced to convert from legacy classes to the newer API objects**.
+
+Sample Code Conversion:
+
+````java
+
+    @Test
+    public void testWorkingWithLegacyDateInJava8() {
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:271]: If you use tests in code examples, please include the class so that readers can find the code easier in the GitHub repo. i.e:

```
public class MyTest {
...
  @Test
    public void testWorkingWithLegacyDateInJava8() {
    }
}```

You can leave out any test methods that are not relevant in the current example

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+        Date date = new Date();
+        System.out.println("java.util.Date : " + date);
+        Instant instant = date.toInstant();
+        System.out.println("Convert java.util.Date to Instant : " + instant);
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:275]: Please use assertions instead of `System.out.println()`, because the reader would have to guess what comes out of a `System.out.println()` (or scroll down to an "output" section)

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+
+## Advantages of the new DateTime API
+- Operations such as formatting, parsing, timezone conversions can be easily performed.
+- Exception handling with classes `java.time.DateTimeException`, `java.time.zone.ZoneRulesException` are well-detailed and easy to comprehend.
+- All classes are immutable making them thread-safe.
+- Each of the classes provides a variety of utility methods that help compute, extract, modify date-time information thus catering to most common use cases.
+- Additional complex date computations are available in conjunction with `java.time.Temporal` package, `java.time.Period` and `java.time.Duration` classes.
+- Methods are added to the legacy APIs to convert objects to `java.time.Instant` and let the legacy code use the newer APIs.
+
+## Dealing with timezones in a SpringBoot application
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:322]: Please check all section titles for TitleCase (https://titlecase.com)

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+
+In this section, we will take a look at how to handle timezones when working with Spring Boot and JPA.
+
+### Introduction to a sample SpringBoot application
+
+For demonstration purposes, we will use [this application](https://github.com/thombergs/code-examples/tree/master/spring-boot/timezones/SpringWebApplication) to look at how timezone conversions apply. This application is a Spring Boot application with MySql as the underlying database.
+First, let's look at the database.
+
+**According to Oracle official documentation**:
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:331]: What documentation? Java? MySQL? Can you link it?

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+    `offset_datetime` TIMESTAMP NULL,
+    `zoned_datetime` TIMESTAMP NULL,
+    `created_at` TIMESTAMP NOT NULL,
+
+    PRIMARY KEY (`id`));
+````
+
+The corresponding JPA entity is as below:
+````java
+    @Entity
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:378]: Please check the formatting of all code snippets.

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+    @Column(name = "offset_datetime")
+    private OffsetDateTime offsetDateTime;
+
+    @Column(name = "zoned_datetime")
+    private ZonedDateTime zonedDateTime;
+
+    @Column(name = "created_at", nullable = false, updatable = false)
+    @CreationTimestamp
+    private OffsetDateTime createdAt;
+
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:413]: Missing `}`?

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+````
+
+### Understanding the server timezone setup
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/TZ.JPG" %}}
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:418]: Please make this a table in the article (again, this is more readable and more inclusive for people with bad eyesight)

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 16]
```diff
@@ -0,0 +1,516 @@
...
+
+As seen, the database should store the timestamp in UTC (+00:00). We will run our spring boot application to the custom timezones **Europe/London** and **Europe/Berlin**.
+Since these timezones have offset +01:00 and +02:00 respectively, we can easily compare the timestamps stored and retrieved.
+
+To start the Spring application in Europe/London timezone we specify the timezone in the arguments as :
+````text
+    mvnw clean verify spring-boot:run -Dspring-boot.run.jvmArguments="-Duser.timezone=Europe/London"
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:425]: To make CLU commands better readable, you can spread them on multiple lines by adding `\` at the end of the line ... they are still copy & pasteable:

```
mvnw clean verify spring-boot:run \
  -Dspring-boot.run.jvmArguments="-Duser.timezone=Europe/London"
```

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 17]
```diff
@@ -0,0 +1,516 @@
...
+````
+
+Similarly, to start the application in Europe/Berlin timezone we would specify the arguments as :
+````text
+    mvnw clean verify spring-boot:run -Dspring-boot.run.jvmArguments="-Duser.timezone=Europe/Berlin"
+````
+
+We have configured two endpoints in the controller class:
+- `http://localhost:8083/app/v1/timezones/default` stores the current date/time in the jvm arguments specified timezone.
+- `http://localhost:8083/app/v1/timezones/dst` stores a custom date/time in the jvm arguments specified timezone.
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:435]: I don't understand what this has to do with DST (the name of the endpoint)?

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 18]
```diff
@@ -0,0 +1,516 @@
...
+{{% info title="`Daylight Savings Time`" %}}
+As on 8th September 2022, both the timezones **Europe/London** and **Europe/Berlin** are on DST. Their corresponding timezones are **British Summer Time (BST) (UTC+1)** and **Central European Summer Time (CEST) (UTC+2)**.
+Post 30th October 2022, the DST will end and they will be back to **Greenwich Mean Time (GMT) (UTC)** and **Central European Time (CET) (UTC+1)** respectively.
+{{% /info %}}
+
+### Comparing results from both timezones
+
+For Europe/London at the current date/time :
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/spring_tz_1.JPG" %}}
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:446]: It's not quite clear from these 2 images what you're doing. Can you explain a bit more? Also, please convert the images to code examples (also the next images of the same kind)

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 19]
```diff
@@ -0,0 +1,516 @@
...
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/db_tz_1.JPG" %}}
+
+On comparison, we can make note of the following points:
+- `VARCHAR` representation of date in the database is not preferred, since it stores the date in the format it was sent. This could result in inconsistent date formats and make it difficult to convert back in the application.
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:450]: In the items of this list, can you refer to the database columns (from the previous image) you are talking about? Otherwise the reader has to guess which one you are talking about.

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 20]
```diff
@@ -0,0 +1,516 @@
...
+With this set, `OffsetDateTime.now(clock)` will always return the same time.
+
+To always apply the default system timezone, we could use:
+````java
+Clock.systemDefaultZone();
+````
+
+By setting the clock parameter, testing the same application in different timezones, with or without DST becomes much easier.
+
+## Conclusion
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:514]: I'm missing a short section "When and how to store timezones in the Database?" that lists the options of how to store timezone information in the database and the pros and cons of each and in which use cases they are best. I.e.:

- store UTC in the database and convert to the user's timezone when reading data out of the DB
- store user timezone in the database (may be required for objects like travel itineraries or manifests or similar things)

**Review by thombergs — CHANGES_REQUESTED** [review]
This is a great article! I left a bunch of comments, but they're mostly concerned with format. This will be a really helpful guide for any Java developer!

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md (hunk 18 above)
Comment (ranjanih) [reply content/blog/2022/2022-08-30-handling-timezones-in-spring.md:445]: Added some additional context here. The images added here are solely to indicate how the Postman output and the Mysql output looks like when we run the application. I have replaced other places when code and output was copied from intellij. However, in this context i think it would make sense to leave the images. Let me know what you think

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 21]
```diff
@@ -0,0 +1,635 @@
...
+
+````java
+@Test
+    public void testCurrentDate() {
+        Date now = new Date();
+        System.out.println(now);
+        Date before = new Date(1661832030000L);
+        System.out.println(before);
+    }
+
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:46]: I see that you have converted some of the code snippets into unit tests, but kept the `System.out.println()`. This way, the unit test doesn't provide any benefit for the reader. As a reader, I still have to switch between the code snippet and the "output" text below it to understand what the code does.

The idea was to replace the `System.out.println()`s with assertions so I can read the result from the code itself. I.e. something like this:

```
@Test
public void testCurrentDate() {
    Date now = new Date();
    assertThat(date).hasDayOfMonth(1)
    assertThat(date).hasMonth(0)
    ...
   Date before = new Date(...);
   assertThat(before)....
}
```

This way, it's clear from a single glance what the code is doing. Can you please convert all the code examples to this format? It will make the article so much more readable.

Instead of using `new Date()` you should then create a specific date so the test doesn't fail after some time.

Actually, for this first example, it probably makes sense to use a `System.out.println()` to show the timezone difference. For most of the other examples, however, a test with assertions would be much more readable.

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 22]
```diff
@@ -0,0 +1,635 @@
...
+````
+
+### ZonedDateTime
+
+**[java.time.ZonedDateTime](https://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html)** is an immutable representation of date, time and timezone. It automatically handles Daylight Saving Time(DST) clock changes via the `java.time.ZoneId` which internally resolves the zone offset.
+
+Sample conversion example:
+
+````java
+ZonedDateTime currentZoneDateTime = ZonedDateTime.now();
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:279]: There are quite a few code examples that have not been converted to unit tests.

**Review by thombergs — CHANGES_REQUESTED** [review]
@ranjanih it seems that not all of my comments have been actioned. There are still a lot of code examples that use `System.out.println()`, the titles are not titleCase, and some screenshots are not converted to tables. Did you maybe forget to push some changes? Please have another look.

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md (hunk 21 above)
Comment (ranjanih) [reply content/blog/2022/2022-08-30-handling-timezones-in-spring.md:45]: Understood. I have addressed them to include assertions wherever applicable. Only the "Comparing timezone results" sections has relevant screenshots. All other sections have code added within the article. Also rectified the casings in some of the titles

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md (hunk 22 above)
Comment (ranjanih) [reply content/blog/2022/2022-08-30-handling-timezones-in-spring.md]: Unit tests are added for all date/time classes

File: content/blog/2022/2022-08-30-handling-timezones-in-spring.md [hunk 23]
```diff
@@ -0,0 +1,798 @@
...
+
+```java
+public class DeprecatedExamples {
+  @Test
+  public void testCurrentDate() {
+    Date now = new Date();
+    Date before = new Date(1661832030000L);
+    assertThat(now).isAfter(before);
+  }
+}
```
Comment (thombergs) [inline content/blog/2022/2022-08-30-handling-timezones-in-spring.md:55]: Please go through the code examples one more time and add the class to the code like I did in this code snippet and the next few snippets. Please also check the code examples in the [preview](https://deploy-preview-502--reflectoring.netlify.app/spring-timezones/) and properly format it and introduce line breaks where the code would stick out of the right side and create a horizontal scroll bar) ... it helps to make indentation with 2 spaces.

Comment (thombergs) [reply content/blog/2022/2022-08-30-handling-timezones-in-spring.md:55]: Sorry, I think my last edit broke the formatting of the code examples a bit, because I had IntelliJ format it for me. Once you have added the class to a code example, however, and then let IntelliJ format it for you (with tabs set to 2 spaces), then IntelliJ does all the work for you (IntelliJ can't format without the wrapping class, however ...).

**Review by thombergs — CHANGES_REQUESTED** [review]
@ranjani2412 I formatted the first few code examples in the way they are best readable, please have a look and copy that style for the other code examples. There are also still some other comments that aren't actioned, yet.
